crawl-ref/source/initfile.cc

```diff
@@ -2067,25 +2079,58 @@ newgame_def read_startup_prefs()
  */
 void game_options::write_prefs(FILE *f)
 {
-    // TODO: generalize, probably some polymorphic functions on GameOption
-    // classes. Not worth doing until more stuff is serialized though...
-    fprintf(f, "default_manual_training = %s\n",
-                        default_manual_training ? "yes" : "no");
-    fprintf(f, "quiver_menu_focus = %s\n",
-                        quiver_menu_focus ? "true" : "false");
+    const auto list =
+    {
+        "default_manual_training", "quiver_menu_focus",
 #ifdef USE_TILE_WEB
-    fprintf(f, "action_panel_orientation = %s\n",
-                        action_panel_orientation.c_str());
-    fprintf(f, "action_panel_show = %s\n",
-                        action_panel_show ? "yes" : "no");
-    fprintf(f, "action_panel_scale = %d\n", action_panel_scale);
-    fprintf(f, "action_panel_font_size = %d\n", action_panel_font_size);
+        "action_panel_orientation", "action_panel_show",
+        "action_panel_scale", "action_panel_font_size",
 #endif
+    };
+
+    for (const auto key : list)
+    {
+        ASSERT(map_find(options_by_name, key));
+        const GameOption &option = **map_find(options_by_name, key);
+        fprintf(f, "%s = %s\n", option.name().c_str(), option.str().c_str());
+    }
+
     // TODO: this variable is extremely coarse, maybe something better? Per
     // opts setting? comparison of serializable values like for newgame_def?
     prefs_dirty = false;
 }
 
+// Save all preferences to the default preference file.
+static void _save_preferences()
+{
+    string fn = find_crawlrc();
+    if (!yesno(make_stringf("Save to '%s'?", fn.c_str()).c_str(), true, 'y'))
+        return;
+
+    const char *mons[12] = { "Jan", "Feb", "Mar", "Apr", "May", "June",
+                             "July", "Aug", "Sept", "Oct", "Nov", "Dec" };
+    FILE *f = fopen_u(fn.c_str(), "a");
+    if (f)
+    {
+        auto date0 = time(nullptr);
+        auto date = TIME_FN(&date0);
+        fprintf(f, "\n# Options saved for %s at %02d:%02d:%02d on %d %s %d.\n",
+                you.your_name.c_str(), date->tm_hour, date->tm_min,
+                date->tm_sec, date->tm_mday, mons[date->tm_mon],
+                1900+date->tm_year);
+        for (auto g : Options.get_sorted_options())
+        {
+            if (!g->name().empty()) // not GameOptionHeading
+                fprintf(f, "%s = %s\n", g->name().c_str(), g->str().c_str());
+        }
```

Comment:
Can delete these braces, right? (If you run `util/unbrace`, won't it do that automatically?)

Reply:
I don't think `unbrace` is quite that clever. I've removed them now.